v3/src/models/document/document-content.ts

```diff
@@ -0,0 +1,199 @@
+    },
+    get rowCount() {
+      return self.rowOrder.length
+    },
+    getRow(rowId: string): ITileRowModelUnion | undefined {
+      return self.rowMap.get(rowId)
+    },
+    getRowByIndex(index: number): ITileRowModelUnion | undefined {
+      return index < self.rowOrder.length ? self.rowMap.get(self.rowOrder[index]) : undefined
+    },
+    getRowIndex(rowId: string) {
+      return self.rowOrder.findIndex(_rowId => _rowId === rowId)
+    },
+    findRowIdContainingTile(tileId: string) {
+      return self.rowOrder.find(rowId => self.rowMap.get(rowId)?.hasTile(tileId))
+    },
+    numTilesInRow(rowId: string) {
+      const row = self.rowMap.get(rowId)
+      return row?.tileCount ?? 0
+    },
+    isPlaceholderRow(row: ITileRowModelUnion) {
+      // Note that more than one placeholder tile in a row shouldn't happen
+      // in theory, but it has been known to happen as a result of bugs.
+      return row.tileIds.every(tileId => {
+        const tile = tileId ? self.tileMap.get(tileId) : undefined
+        return isPlaceholderTile(tile)
+      })
+    },
+    get indexOfLastVisibleRow() {
+      // returns last visible row or last row
+      if (!self.rowOrder.length) return -1
+      const lastVisibleRowId = self.visibleRows.length
+                                ? self.visibleRows[self.visibleRows.length - 1]
+                                : self.rowOrder[self.rowOrder.length - 1]
+      return  self.rowOrder.indexOf(lastVisibleRowId)
+    },
+    getFirstSharedModelByType<IT extends typeof SharedModel>(
+      modelType: IT, tileId?: string): IT["Type"] | undefined {
+      const sharedModelEntries = Array.from(self.sharedModelMap.values())
+      // Even if we use a snapshotProcessor generated type, getType will return the original type
+      const firstEntry = sharedModelEntries.find(entry =>
+        (getType(entry.sharedModel) === modelType) &&
+        (!tileId || !!entry.tiles.find(tile => tileId === tile.id)))
+      return firstEntry?.sharedModel
+    },
+    getSharedModelsByType<IT extends typeof SharedModel>(type: string): IT["Type"][] {
+      const sharedModelEntries = Array.from(self.sharedModelMap.values())
+      return sharedModelEntries.map(entry => entry.sharedModel).filter(model => model.type === type)
+    }
+  }))
+  .views(self => ({
+    get defaultInsertRow() {
+      // next tile comes after the last visible row with content
+      for (let i = self.indexOfLastVisibleRow; i >= 0; --i) {
+        const row = self.getRowByIndex(i)
+        if (row?.acceptDefaultInsert) return row
+        if (row && !row.isSectionHeader && !self.isPlaceholderRow(row)) {
+          return i + 1
+        }
+      }
+      // if no tiles have content, insert after the first non-header row
+      for (let i = 0; i < self.rowCount; ++i) {
+        const row = self.getRowByIndex(i)
+        if (row?.acceptDefaultInsert) return row
+        if (row && !row.isSectionHeader) {
+          return i + 1
```

Comment:
I'm not totally sure, but it seems like this should just be `return i` like it used to be.

Reply:
I came to the conclusion that this was a bug in the original implementation. The two `for` loops are completely analogous and in both cases they are documented as returning the index of the row _after_ the one that satisfies certain conditions, so I think it should be `i + 1` in both cases.